### PR TITLE
Implement Drawable::Calc_Physics_Xform_Treads

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/client/draw/drawable.cpp [hunk 1]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+
+    float accel_pitch_limit = locomotor->Get_Accel_Pitch_Limit();
+    float deacel_pitch_limit = locomotor->Get_Deaccel_Pitch_Limit();
+    float pitch_stiffness = locomotor->Get_Pitch_Stiffness();
+    float roll_stiffness = locomotor->Get_Roll_Stiffness();
+    float pitch_damping = locomotor->Get_Pitch_Damping();
+    float roll_damping = locomotor->Get_Roll_Damping();
+    float forward_accel_coef = locomotor->Get_Foward_Accel_Coef();
+    float lateral_accel_coef = locomotor->Get_Lateral_Accel_Coef();
+    float uniform_axial_damping = locomotor->Get_Uniform_Axial_Damping();
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp:2417]: All these variables and calls should be put further down the scopes to where they are used. It makes no sense to have them before `object != nullptr` and similar, where they are potentially not used at all.

File: src/game/client/draw/drawable.cpp [hunk 2]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+{
+    if (m_drawableLocoInfo == nullptr) {
+        m_drawableLocoInfo = new DrawableLocoInfo();
+    }
+
+    float f1 = 0.8f;
+    float f2 = 0.5;
+    float f3 = 0.024543693;
+    float f4 = 5.0;
+    float f5 = 0.5;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: All of these values above should be put in place where they are used. That makes is easier to look at it. Most of them are just used once anyway.

File: src/game/client/draw/drawable.cpp [hunk 3]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+
+                        if (f9 >= 0.0f) {
+                            if (f9 > 1.0f) {
+                                f9 = 1.0f;
+                            }
+                        } else {
+                            f9 = 0.0f;
+                        }
+
+                        float random = (velocity.x * velocity.x + velocity.y * velocity.y) * f4;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `max_velocity`

File: src/game/client/draw/drawable.cpp [hunk 4]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                g_theTerrainLogic->Get_Layer_Height(position->x, position->y, layer, &n, true);
+                float overlap_z = 0.0f;
+                Object *overlap = g_theGameLogic->Find_Object_By_ID(physics->Get_Current_Overlap());
+
+                if (overlap != nullptr && overlap->Is_KindOf(KINDOF_SHRUBBERY)) {
+                    overlap = nullptr;
+                }
+
+                if (overlap != nullptr) {
+                    const Coord3D *pos2 = overlap->Get_Position();
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `overlap_pos`

File: src/game/client/draw/drawable.cpp [hunk 5]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                    float posx = pos2->x - position->x;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `overlap_offset_x`

File: src/game/client/draw/drawable.cpp [hunk 6]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                    float posy = pos2->y - position->y;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `overlap_offset_y`

File: src/game/client/draw/drawable.cpp [hunk 7]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                    float f6 = GameMath::Square(posy) + GameMath::Square(posx);
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `overlap_dist_sqr`

File: src/game/client/draw/drawable.cpp [hunk 8]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                    float radius = Get_Drawable_Geometry_Info().Get_Bounding_Circle_Radius();
+                    float radius2 = overlap->Get_Geometry_Info().Get_Bounding_Circle_Radius();
+                    float f7 = (radius2 + radius) * f1;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: Maybe `dist_threshold`

File: src/game/client/draw/drawable.cpp [hunk 9]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+
+                    if (GameMath::Square(f7) > f6) {
+                        float f8 = GameMath::Sqrt(f6);
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: `overlap_dist`

File: src/game/client/draw/drawable.cpp [hunk 10]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+                        float f9 = 1.0f - f8 / f7;
```
Comment (xezon) [inline src/game/client/draw/drawable.cpp]: Maybe `overlap_speed`

**Comment by xezon** [comment]
Treads movement looks good in game.

File: src/game/client/draw/drawable.cpp (hunk 2 above)
Comment (tomsons26) [reply src/game/client/draw/drawable.cpp]: these need to be f

File: src/game/client/draw/drawable.cpp [hunk 11]
```diff
@@ -2393,3 +2393,191 @@ void Drawable::Calc_Physics_Xform_Wheels(Locomotor const *locomotor, PhysicsXfor
...
+
+                        if (f9 >= 0.0f) {
```
Comment (tomsons26) [inline src/game/client/draw/drawable.cpp]: flip

File: src/game/client/draw/drawable.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: Added the f.
Not moving the variables, want to keep it matching the pseudo to make it easier to follow.

File: src/game/client/draw/drawable.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp:2417]: Not moving the variables, want to keep it matching the pseudo to make it easier to follow.

File: src/game/client/draw/drawable.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

File: src/game/client/draw/drawable.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/client/draw/drawable.cpp]: fixed.

**Review by tomsons26 — APPROVED** [review]

**Pull request merged**